src/Esprima/ArrayList.cs

```diff
@@ -120,7 +150,62 @@ public int Count
         }
     }
 
-    internal void AddRange<TSource>(ArrayList<TSource> list) where TSource : T
+    public T this[int index]
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        get
+        {
+            AssertUnchanged();
+
+            // Following trick can reduce the range check by one
+            if ((uint) index < (uint) _count)
+            {
+                return _items![index];
+            }
+
+            return ThrowArgumentOutOfRangeException<T>(nameof(index), index, null);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        set
+        {
+            AssertUnchanged();
+
+            // Following trick can reduce the range check by one
+            if ((uint) index < (uint) _count)
```

Comment:
Same here, local ref to array and check. Length instead of count

Reply:
I'm afraid that nothing's left to improve here. This piece of code was obviously borrowed from [here](https://github.com/dotnet/runtime/blob/v6.0.8/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L143).

We can't really check `items.Length` because it's not necessarily equal to `_count` (may be greater than that). I think the comment about the trick means that using the `uint` cast we need one comparison instead of two (i.e. `index < 0 || index >= _count`).